src/gui/sdlmain.cpp

```diff
@@ -3307,16 +3196,44 @@ static SDL_Rect calc_viewport_fit(const int win_w, const int win_h)
 	// limit the window to the user's desired viewport, if configured
 	const auto [lwin_w, lwin_h] = restrict_to_viewport_resolution(win_w, win_h);
 
-	// calculate the aspect ratios of the draw buffer and the limited window
+	// calculate the aspect ratio of the draw buffer
 	const auto draw_aspect = (sdl.draw.width * sdl.draw.scalex) /
 	                         (sdl.draw.height * sdl.draw.scaley);
-	const auto lwin_aspect = static_cast<double>(lwin_w) / lwin_h;
 
-	// calculate the viewport contingent on the aspect ratio of the limited
-	// window versus draw buffer
-	const auto lwin_is_wider = lwin_aspect > draw_aspect;
-	const auto view_w = lwin_is_wider ? iround(lwin_h * draw_aspect) : lwin_w;
-	const auto view_h = lwin_is_wider ? lwin_h : iround(lwin_w / draw_aspect);
+	int view_w = 0;
+	int view_h = 0;
+	switch (sdl.integer_scale_mode) {
+	case INTEGER_SCALE_MODE::OFF: {
+		// calculate the viewport contingent on the aspect ratio of the
+		// limited window versus draw buffer
+		const auto lwin_aspect   = static_cast<double>(lwin_w) / lwin_h;
+		const auto lwin_is_wider = lwin_aspect > draw_aspect;
+		view_w = lwin_is_wider ? iround(lwin_h * draw_aspect) : lwin_w;
+		view_h = lwin_is_wider ? lwin_h : iround(lwin_w / draw_aspect);
+		break;
+	}
+	case INTEGER_SCALE_MODE::HORIZONTAL: {
+		const auto integer_scale_factor = std::max(
+		        std::min(lwin_w / sdl.draw.width,
+		                 static_cast<int>(lwin_h / (sdl.draw.height *
+		                                            sdl.draw.pixel_aspect))),
+		        1);
+		view_w = sdl.draw.width * integer_scale_factor;
+		view_h = iround(sdl.draw.height * integer_scale_factor *
+		                sdl.draw.pixel_aspect);
+		break;
+	}
+	case INTEGER_SCALE_MODE::VERTICAL: {
+		const auto integer_scale_factor = std::max(
+		        std::min(lwin_h / sdl.draw.height,
+		                 static_cast<int>(lwin_w / (sdl.draw.height *
+		                                            draw_aspect))),
+		        1);
+		view_w = iround(sdl.draw.height * integer_scale_factor * draw_aspect);
+		view_h = sdl.draw.height * integer_scale_factor;
```

Comment:
These are quite hard to follow, I spent like 2-3 minutes on it and gave up... I could figure it out eventually, but we really should optimise for readability by default. This is not performance intensive code, and even if it was, the compiler can usually optimise it well enough...

Can break this up into multiple statements with individual variable assignments with descriptive names instead of giving us the "final closed form"? Like in maths class; the teacher might see the final "closed form solution" already, but you need to explicitly tell the kids all the 10 individual derivation steps to see the solution. Well, we're the kids here 😄 

As an added benefit, that serves as documentation as well, instead of just giving people the final formula, then wishing them a nice day 😎

Reply:
Replaced `std.max()` with a separate expression and added comments. Formatter has... interesting views on how these lines should look, so a bit more refinement might be needed if you find that exceptionally ugly.